From aef5982132e77ed2aa124488d116e9cbc75bacf7 Mon Sep 17 00:00:00 2001
From: He Sheng <hesheng@wxiat.com>
Date: Tue, 7 Jun 2022 17:42:49 +0800
Subject: [PATCH] sw64: push and pop kernel stack with ldi instruction
Patch-mainline: NO, OTHERS
References: openEuler-22.03-LTS
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I56OLG

--------------------------------

addl and subl only support 8-bit immediate value, which is not
suitable for us to extend struct pt_regs later.

Signed-off-by: He Sheng <hesheng@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/kernel/entry.S | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/arch/sw_64/kernel/entry.S b/arch/sw_64/kernel/entry.S
index c12487e52141..6c40d2015439 100644
--- a/arch/sw_64/kernel/entry.S
+++ b/arch/sw_64/kernel/entry.S
@@ -22,7 +22,7 @@
  */
 
 #define SAVE_ALL				\
-	subl	$sp, PT_REGS_PS, $sp;		\
+	ldi	$sp, -PT_REGS_PS($sp);		\
 	stl	$0, PT_REGS_R0($sp);		\
 	stl	$1, PT_REGS_R1($sp);		\
 	stl	$2, PT_REGS_R2($sp);		\
@@ -66,7 +66,7 @@
 	ldl	$26, PT_REGS_R26($sp);		\
 	ldl	$27, PT_REGS_R27($sp);		\
 	ldl	$28, PT_REGS_R28($sp);		\
-	addl	$sp, PT_REGS_PS, $sp
+	ldi	$sp, PT_REGS_PS($sp)
 
 /*
  * Non-syscall kernel entry points.
-- 
2.34.1

